compiler/ilgen/VirtualMachineOperandStack.cpp

```diff
@@ -0,0 +1,197 @@
+VirtualMachineOperandStack::MergeInto(OMR::VirtualMachineOperandStack *other, TR::IlBuilder *b)
+   {
+   TR_ASSERT(_stackTop == other->_stackTop, "stacks are not same size");
+   for (int32_t i=_stackTop;i >= 0;i--)
+      {
+      // only need to do something if the two entries aren't already the same
+      if (other->_stack[i]->getCPIndex() != _stack[i]->getCPIndex())
+         {
+         // what if types don't match? could use ConvertTo, but seems...arbitrary
+         // nobody *should* design bytecode set where corresponding elements of stacks from
+         // two incoming control flow edges can have different primitive types. objects, sure
+         // but not primitive types! Expecting to be disappointed here some day...
+         TR_ASSERT(_stack[i]->getSymbol()->getDataType() == other->_stack[i]->getSymbol()->getDataType(), "disappointing stack merge: primitive type mismatch at same depth stack elements");
+         b->Store(other->_stack[i]->getSymbol()->getAutoSymbol()->getName(),
+               _stack[i]);
+         }
+      }
+   }
+
+// Allocate a new operand stack and copy everything in this state
+// If VirtualMachineOperandStack is subclassed, this function *must* also be implemented in the subclass!
+VirtualMachineState *
+VirtualMachineOperandStack::MakeCopy()
+   {
+   VirtualMachineOperandStack *copy = (VirtualMachineOperandStack *) TR::comp()->trMemory()->allocateHeapMemory(sizeof(VirtualMachineOperandStack));
+   new (copy) VirtualMachineOperandStack(this);
+
+   return copy;
+   }
+
+void
+VirtualMachineOperandStack::Push(TR::IlBuilder *b, TR::IlValue *value)
+   {
+   checkSize();
+   _stack[++_stackTop] = value;
+   _stackTopRegister->Adjust(b, +_pushAmount);
+   }
+
+TR::IlValue *
+VirtualMachineOperandStack::Top()
+   {
+   TR_ASSERT(_stackTop >= 0, "no top: stack empty");
+   return _stack[_stackTop];
+   }
+
+TR::IlValue *
+VirtualMachineOperandStack::Pop(TR::IlBuilder *b)
+   {
+   TR_ASSERT(_stackTop >= 0, "stack underflow");
+   _stackTopRegister->Adjust(b, -_pushAmount);
+   return _stack[_stackTop--];
+   }
+
+TR::IlValue *
+VirtualMachineOperandStack::Pick(int32_t depth)
+   {
+   TR_ASSERT(_stackTop >= depth, "pick request exceeds stack depth");
+   return _stack[_stackTop - depth];
+   }
+
+void
+VirtualMachineOperandStack::Drop(TR::IlBuilder *b, int32_t depth)
+   {
+   TR_ASSERT(_stackTop >= depth-1, "stack underflow");
+   _stackTop-=depth;
+   _stackTopRegister->Adjust(b, -depth * _pushAmount);
+   }
+
+void
+VirtualMachineOperandStack::Dup(TR::IlBuilder *b)
+   {
+   TR_ASSERT(_stackTop >= 0, "cannot dup: stack empty");
+   TR::IlValue *top = _stack[_stackTop];
+   checkSize();
+   _stack[++_stackTop] = top;
+   _stackTopRegister->Adjust(b, +_pushAmount);
+   }
+
+void
+VirtualMachineOperandStack::copyTo(VirtualMachineOperandStack *copy)
+   {
+   }
+
+void
+VirtualMachineOperandStack::checkSize()
+   {
+   if (_stackTop == _stackMax)
+      grow();
+   }
+
+void
+VirtualMachineOperandStack::grow()
```

Comment:
Did you consider a grow method that takes an "increment amount" as a parm as opposed to having a "standard" increment amount applied inside of grow() ?